---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

# enRUVseq

<!-- badges: start -->

<!-- badges: end -->

The goal of `enRUVseq` is to perform normalization on RNA-seq including
enrichment (NAD-RNA-seq) using spike-in.

The main functions for normalizing enrichment variation between samples
were inspired by [RUVSeq](https://github.com/drisso/RUVSeq).

## Installation

You can install the development version of enRUVseq from [GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("thereallda/enRUVseq")
```

## Quick tour 

```{r example}
library(enRUVseq)
library(tidyverse)
library(edgeR)
```

### Load data

Metadata including sample information. 

```{r}
meta <- read.csv('data-raw/metadata_enRUVg.csv', comment.char = '#')
meta
```

NAD-RNA-Sequencing data, with genes in rows and samples in columns. 

```{r}
counts.df <- read.csv('data-raw/Counts_enRUVg.csv', row.names = 1)
counts.df[1:3,]
```

### Filtering low-expressed genes 

```{r}
keep <- filterByExpr(counts.df, group = meta$condition)
counts.keep <- counts.df[keep,]
dim(counts.df); dim(counts.keep)
```

### Perform RUV normalization

Both RUVg and RUVs return the following:

 - `dataNorm`: the normalized count matrix of shape n x p, where n is the number of samples and p is the number of features. 
 - `adjustFactor`: the adjusting factor for removing unwanted variation (n x k). 
 - `alpha`: the nuisance parameters (k x p). 

- RUVg

```{r}
spikeInPrefix <- '^FB' # prefix of spike-in gene id
control.idx <- grep(spikeInPrefix, rownames(counts.keep), value = TRUE)
set.ruvg <- normRUV(counts.keep, 
                    control.idx = control.idx,
                    method = 'RUVg')
str(set.ruvg)
```

- RUVs 

When using `RUVs`, replicated samples needed to be specified with a matrix `sc.idx`. 

`sc.idx` A numeric matrix specifying the replicate samples for which to 
compute the count differences used to estimate the factors of unwanted variation.

```{r}
spikeInPrefix <- '^FB' # prefix of spike-in gene id
control.idx <- grep(spikeInPrefix, rownames(counts.keep), value = TRUE)
sc.idx <-  t(sapply(unique(meta$condition), function(i) grep(i, meta$condition)))
sc.idx 
```

```{r}
set.ruvs <- normRUV(counts.keep, 
                    control.idx = control.idx,
                    sc.idx = sc.idx,
                    method = 'RUVs')
str(set.ruvs)
```

### Differential analysis with edgeR

One can use `edgeR` to perform differential analysis with adjusting factors estimated from RUV. 

First, specifying the contrasts

```{r}
contrast_df <- data.frame(Group1 = unique(grep("Enrich", meta$condition, value = TRUE)),
                          Group2 = unique(grep("Input", meta$condition, value = TRUE)))
contrast_df
```

Then, perform `edgeR` procedure with non-spike-in counts

```{r}
counts.nsp <- counts.keep[!rownames(counts.keep) %in% control.idx, ]
de.nsp1 <- edgeRDE(counts.nsp, 
            group = meta$condition,
            contrast.df = contrast_df,
            adjust.factors = set.ruvg$adjustFactor) 

str(de.nsp1)
```

`edgeRDE` return a list of differential analysis objects, including:
- `de.obj`: differential analysis object
- `res.ls`: list of differential analysis unfiltered-results tables
- `res.sig.ls`: list of filtered results tables, default `logFC >= 1 & FDR < 0.05`

Similarly, one can perform with adjusting factors from RUVs

```{r}
de.nsp2 <- edgeRDE(counts.nsp, 
            group = meta$condition,
            contrast.df = contrast_df,
            adjust.factors = set.ruvs$adjustFactor) 

str(de.nsp2)
```

### Visualization

PCA before and after normalization

```{r}
library(patchwork)
samples_name <- paste(meta$condition, meta$replicate, sep = '.')
# raw counts
p1 <- ggPCA(log2(counts.nsp + 1), labels = samples_name, vst.norm = FALSE) + ggtitle('Raw')
# RUV normalized counts
counts.nsp.ruv <- set.ruvg$dataNorm[!rownames(set.ruvg$dataNorm) %in% control.idx,]
p2 <- ggPCA(log2(counts.nsp.ruv + 1), labels = samples_name, vst.norm = FALSE) + ggtitle('RUV')

p1 + p2
```

The global difference of fold-change of significant DEGs

```{r}
# reduce a list of DE tables into one data frame
df.nsp <- reduceRes(de.nsp1$res.sig.ls, fc.col = 'logFC') # fc.col specify the name of log-fold change column
head(df.nsp)
```
```{r}
# Simplify group name
df.nsp$Group <- gsub('\\..*', '', df.nsp$Group)
df.nsp$Group <- factor(df.nsp$Group, levels = unique(df.nsp$Group))
# visualization with box vilion plot
BetweenStatPlot(df.nsp, 
                x = 'Group',
                y = 'logFC',
                color = 'Group')
```

